server/lib/api/users.js

```diff
@@ -97,3 +106,33 @@ async function resetPassword(ctx, next) {
     ctx.status = 204
   })
 }
+
+async function verifyEmail(ctx, next) {
+  const { id } = ctx.params
+  const { code } = ctx.query
+  const { email } = ctx.request.body
+
+  if (!id || !code || !isValidEmail(email)) {
+    throw new httpErrors.BadRequest('Invalid parameters')
+  }
+
+  const verifiedEmail = await useEmailVerificationCode(id, email, code)
+  if (!verifiedEmail) {
+    throw new httpErrors.BadRequest('Email verification code is invalid')
+  }
+  ctx.body = verifiedEmail
+}
+
+async function resendVerificationEmail(ctx, next) {
```

Comment:
This needs to be rate-limited very heavily to prevent using it as a spam vector. We may want to consider some maximum amount of times we'd send email to a particular address without verification (e.g. after N attempts you are required to specify a different email address, rather than being able to re-send it).